almond/enginemanagerclient.js

```diff
@@ -16,11 +16,25 @@ const rpc = require('transparent-rpc');
 const sockaddr = require('sockaddr');
 
 const JsonDatagramSocket = require('../util/json_datagram_socket');
+const userToShardId = require('./shard');
 
 const Config = require('../config');
 
 var _instance;
 
+function connectToMaster(shardId) {
+    const shard = Config.THINGENGINE_MANAGER_ADDRESS[shardId];
+
+    const socket = new net.Socket();
+    socket.connect(sockaddr(shard));
+
+    const jsonSocket = new JsonDatagramSocket(socket, socket, 'utf8');
```

Comment:
do we care about the order of messages? we probably need that in your editing project right?

Reply:
Yes we care about the order of messages - but you're mixing levels if you worry about user visible almond messages here!
Messages here means something like:
{"control":"call","target":1234558,"method":"getAllApps","args":[],"id":7}
{"control":"reply","id":7,"result":[{...}]}

The actual protocol is implemented by transparent-rpc, a tiny RPC library I wrote some time ago that operates over any object-mode stream capable of sending JSON blobs.

In any case, jsondatagramsocket is a totally ordered socket; the name datagram just refers to the fact that you send data in discrete blobs of JSON rather than a continuous stream of bytes.